feat: New test package (#158)
Fryuni/inox-tools: docs/src/content/docs/astro-tests.mdx

It offers utilities for setting up environments, running builds, and inspecting renders, ensuring compatibility and performance without the need to tamper with internal Astro logic.

## Compatibility

Relying on internal details of Astro can lead to your code breaking at any time. Those internal pieces can move, break or be removed from Astro even on patch versions since they are not part of the public API.

Although this library also utilizes some parts of the Astro code that are meant to be internal, projects using this library for their tests will not break in case those parts are changed. This is ensured by how it built and published, although the source code breaks encapsulation, the code in the published library relies only on Astro's public API.

You can check on your node_modules to see it 😉

## How to install

import { PackageManagers } from 'starlight-package-managers';

<PackageManagers dev pkg="@inox-tools/astro-tests" />

## Fixtures

### `config`

<p>**Type:** `AstroConfig`</p>

> The final config passed to [Astro's programatic CLI entrypoints](https://docs.astro.build/en/reference/cli-reference/#advanced-apis-experimental). This configuration can be overriden on each method call.

Reviewer: ```suggestion
The final config passed to [Astro's programatic CLI entrypoints](https://docs.astro.build/en/reference/cli-reference/#advanced-apis-experimental). This configuration can be overridden for each method call.
```

> Will be automatically passed to the methods below:

Reviewer: ```suggestion
It will automatically be passed to the following methods:
```

- [`startDevServer()`](#startdevserver)
- [`build()`](#build)
- [`preview()`](#preview)
- [`sync()`](#sync)

### `startDevServer`

{(

<p>
	<strong>Type: </strong>
	<code dir="auto" style="padding-right: 0;">
		(inlineConfig:
	</code>
	<a href="https://docs.astro.build/en/reference/cli-reference/#astroinlineconfig">
		<code dir="auto" style="padding-right: 0;">
			AstroInlineConfig
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;">
		) =&gt; Promise&lt;DevServer&gt;
	</code>
</p>
)}

> Starts a dev server at an available port.

Reviewer: ```suggestion
Starts a development server on an available port.
```

> This server can't be running at the same time for thein same fixture as .preview() since they share ports.

Reviewer: ```suggestion
This server cannot run simultaneously with `.preview()` for the same fixture, as they share ports. Ensure `devServer.stop()` is called before the test exits.
```

> Be sure to call devServer.stop() before test exit.

Reviewer: ```suggestion
Ensure `devServer.stop()` is called before the test exits.
```

Equivalent to running [`astro dev`](https://docs.astro.build/en/reference/cli-reference/#astro-dev).

### `build`

{(

<p>
	<strong>Type: </strong>
	<code dir="auto" style="padding-right: 0;">
		(inlineConfig:
	</code>
	<a href="https://docs.astro.build/en/reference/cli-reference/#astroinlineconfig">
		<code dir="auto" style="padding-right: 0;">
			AstroInlineConfig
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;">
		) =&gt; Promise&lt;void&gt;
	</code>
</p>
)}

Builds into current folder (will erase previous build).

Equivalent to running [`astro build`](https://docs.astro.build/en/reference/cli-reference/#astro-build).

### `preview`

{(

<p>
	<strong>Type: </strong>
	<code dir="auto" style="padding-right: 0;">
		(inlineConfig:
	</code>
	<a href="https://docs.astro.build/en/reference/cli-reference/#astroinlineconfig">
		<code dir="auto" style="padding-right: 0;">
			AstroInlineConfig
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;">
		) =&gt; Promise&lt;PreviewServer&gt;
	</code>
</p>
)}

Starts a preview server.

> This server can't be running at the same time for thein same fixture as .dev() since they share ports.

Reviewer: ```suggestion
This server cannot run simultaneously with `.dev()` on the same fixture, as they share ports.
```

> Be sure to call server.stop() before test exit.

Reviewer: ```suggestion
Ensure `server.stop()` is called before the test exits.
```

### `pathExists`

<p>**Type:** `(path: string) => boolean`</p>

Checks whether the given path exists on the build output (`outDir` config).

### `readFile`

<p>
  **Type:** `(path: string) => Promise<string>`
</p>

Read a file from the build (relative to `outDir` config).

### `editFile`

<p>
  **Type:** `(path: string, updater: string | ((content: string) => string)) => Promise<() => void>`
</p>

Edit a file in the fixture directory.

The first parameter is a path relative to the root of the fixture.

The second parameter can be the new content of the file or a function that takes the current content and returns the new content.

This function returns a Promise that resolves to another function. This resolved function can be called to revert the changes.

All changes made with `editFile` are automatically reverted before the [process exits](https://nodejs.org/api/process.html#event-exit).

### `resetAllFiles`

<p>**Type:** `() => void`</p>

Reset all changes made with [`.editFile()`](#editfile)

### `readdir`

<p>
  **Type:** `(path: string) => Promise<string[]>`
</p>

Read a directory from the build output (relative to `outDir` config).

> This is a convenience wrapper around [readdir from Node's FS Promise API](https://nodejs.org/api/fs.html#fspromisesreaddirpath-options).

Reviewer: ```suggestion
This is a convenience wrapper for [readdir from Node's FS Promise API](https://nodejs.org/api/fs.html#fspromisesreaddirpath-options).
```

### `loadNodeAdapterHandler`

{(

<p>
	<strong>Type: </strong>
	<code dir="auto" style="padding-right: 0;">
		() =&gt; Promise&lt;(req:{' '}
	</code>
	<a href="https://nodejs.org/api/http.html#class-httpincomingmessage">
		<code dir="auto" style="padding-left: 0;padding-right: 0;">
			http.IncomingMessage
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;padding-right: 0;">
		, res:{' '}
	</code>
	<a href="https://nodejs.org/api/http.html#class-httpserverresponse">
		<code dir="auto" style="padding-left: 0;padding-right: 0;">
			http.ServerResponse
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;">
		) =&gt; void&gt;
	</code>
</p>
)}

Load the handler for an app built using the [Node Adapter](https://docs.astro.build/en/guides/integrations-guide/node/).

The handler is the same as a listener for the [`request` event](https://nodejs.org/api/http.html#event-request) from Node's native HTTP module.

### `loadTestAdapterApp`

<p>
  **Type:** `() => Promise<TestApp>`
</p>

#### `TestApp`

```ts
type TestApp = {
	render: (req: Request) => Promise<Response>;
	toInternalApp: () => App;
};
```

> A minimal proxy to the underlying Astro App using the test adapter.

Reviewer: ```suggestion
A minimal proxy for the underlying Astro app, provided by the test adapter.
```

:::danger
This class is internal, undocumented and highly unstable. Use it at your own risk.
:::

## Test Adapter

An [Astro Adapter](https://docs.astro.build/en/guides/server-side-rendering/) that exposes the rendering process to be called directly.

It also collects information about the build passed to the adapter to be inspected.

None of the options are required.

### `env`

<p>
  **Type:** `Record<string, string | undefined>`
</p>

Server-side environment variables to be used by [`astro:env`](https://docs.astro.build/en/reference/configuration-reference/#experimentalenv).

### `setRoutes`

{(

<p>
	<strong>Type: </strong>
	<code dir="auto" style="padding-right: 0;">
		(routes:{' '}
	</code>
	<a href="https://docs.astro.build/en/reference/integrations-reference/#routedata-type-reference">
		<code dir="auto" style="padding-left: 0;padding-right: 0;">
			RouteData
		</code>
	</a>
	<code dir="auto" style="padding-left: 0;">
		[]) =&gt; Promise&lt;void&gt;
	</code>
</p>
)}

A callback function that will receive the final value of the project routes.

## Utilities

### No Node checker

> A Vite plugin to ensure no module in the final bundle relies on the built-in Node modules.

Reviewer: ```suggestion
A Vite plugin that ensures no module in the final bundle depends on built-in Node modules.
```

> If the generated bundle contains any reference to a Node module the build will fail.

Reviewer: ```suggestion
If any reference to a Node module is found in the generated bundle, the build will fail.
```
